Pool liquidity calculations

timjrobinson opened the pull request:
- Adds ability to calculate pool liquidity for weighted/metastable pools.
- Adds tokenprovider structure to provide token information where the backend can be replaced with different data sources. 
- Adds example with real pool information to demonstrate how these functions work, using only data the API server already has (pulled from the graph and stored in DynamoDB. 

This needs to be reformatted into the new liquidity concerns structure. Pushing here to Jarek can see the WIP while I'm away at the conference.

gmbronco commented on balancer-js/src/modules/liquidity/liquidity.module.ts:
```diff
@@ -0,0 +1,132 @@
+}
+
+export class Liquidity {
+    constructor(
+        config: BalancerSdkConfig,
+        private pools: PoolProvider,
+        private tokens: TokenProvider,
+        private tokenPrices: TokenPriceProvider
+    ) {}
```
isn't liquidity just a sum of token balance * it's price? how about just returning a token[] from this and running the calculation on it? reason is to remove a hidden dependency on a Pools class.

timjrobinson replied:
We need the pools provider for recursive liquidity calculations, as it loads the subpools, loads the tokens from them, and calculates their liquidity.

gmbronco reviewed (commented):
nice one, i like how data layer is getting in shape. if it's possible, let's kill sor dependency on the data as well and extract the subgraph client for there.

in the meantime i worked on a general pools interface with separated data / model logic: https://gist.github.com/gmbronco/62b060456642533d30cc6ed31ca7b97d

main idea is expanding on a clean data dependencies and makes a space for some data preparation logic on top of raw data. question if the flow would work for other "decoration" use-cases.

gmbronco commented on balancer-js/examples/pools/calculateLiquidity.ts:
```diff
@@ -0,0 +1,114 @@
+            decoratedPool?.totalLiquidity
+        );
+        console.log('---');
+    }
```
```suggestion
    }
    
    process.exit(0);
```

timjrobinson commented on balancer-js/src/modules/liquidity/liquidity.module.ts:
```diff
@@ -0,0 +1,132 @@
+
+                const tokenPrice = await this.tokenPrices.find(
+                    tokenDetails.address
+                );
```
True, not sure what I was thinking there

timjrobinson commented on balancer-js/src/modules/data-providers/pool/sor.provider.ts:
```diff
@@ -0,0 +1,45 @@
+import { SOR, SubgraphPoolBase } from '@balancer-labs/sor';
+import { PoolAttribute, PoolProvider } from './provider.interface';
+
+export class SORPoolProvider implements PoolProvider {
```
That makes sense, I'd like to eliminate SOR as a dependency for this.

timjrobinson commented:
@gmbronco I've tested this in the frontend using current data and it returns the same results as the frontend does now (see https://github.com/balancer-labs/frontend-v2/pull/1950) so is good to review + merge.

gmbronco commented:
Hey @timjrobinson thanks for the update. few issues:
* example examples/pools/calculateLiquidity.ts is crashing:
* does Liquidity need config?
* test/lib/liquidityPools.json -> test/fixtures/liquidityPools.json
* PoolToken vs TokenBalance types, seems like we have a lot of overlap, what's the reason to add a new one?
* isn't `subPoolLiquidity` = BPT price * balance? does it make sense to extract BPT price calculation somewhere, eg: to pricing context?

Side note for future refactoring: Long term we will probably be migrating towards Pool interface as something exposed outside of SDK, something like: Pool.find(id).liquidity

timjrobinson commented:
1. Fixed
2. You're right. It also doesn't need tokens any more. Removed them
3. Fixed
4. You're right, I've consolidated TokenBalance into Token and PoolToken which inherits from that and includes the pool specific information, having Token and TokenBalance was confusing.
5. Yep, have updated code to be clearer that this is what is happening.

gmbronco commented:
something I was thinking about is file naming. right now we are using quite extensive names, what do you think about shortening them up a bit? eg:

```
modules/data-providers/pool/provider.interface.ts -> modules/data-providers/pool/interfaces.ts (we already know we are in providers)
modules/data-providers/pool/static.provider.ts -> modules/data-providers/pool/static.ts
modules/liquidity.module.ts -> modules/liquidity/liquidity.ts, exported from index.ts for convinience
```


anyhow lgtm! cc: @johngrantuk

timjrobinson commented:
> something I was thinking about is file naming. right now we are using quite extensive names, what do you think about shortening them up a bit? eg:

I like this idea. I also like shortening the instantiations of those classes, like changing the variable from `pricingCalculator` to `pricing` (as I did for liquidity).
